src/replication.c

```diff
@@ -2649,6 +2678,12 @@ void freePendingReplDataBuf(void) {
     server.pending_repl_data.len = 0;
 }
 
+void receiveRDBinBioThread(connection *conn) {
+    serverLog(LL_NOTICE, "Replica main thread creating Bio thread to save RDB to disk");
+    connSetReadHandler(conn, NULL);
```

Comment:
What happens to the write handler? Is the main thread supposed to do any writes in the meantime?

Reply:
Good question, it was never set to anything on the replica's side, see:
https://github.com/valkey-io/valkey/blob/unstable/src/replication.c#L3517

^ this is executed during sync handshake. After the sync is done and we enter steady-state we initialize it: https://github.com/valkey-io/valkey/blob/unstable/src/replication.c#L4380

Reply:
@nitaicaro so what about it?

How do we handle the KA notification to the primary during load? 
The way I know it is done in replicationSendNewlineToPrimary, but now it has no connection? 
also if it did, I do not think it is good to have different threads do the read and write. probably will work in most implementations due to the full duplex implementation, but some L5 parts might break (like openssl)

Reply:
We don't have to do it when _receiving_ the RDB - the Bio thread has the connection and it continuously reads from it, so we don't time out.

When _loading_ the RDB, we don't read from the primary's connection, and the primary could think that the replica timed out, so it uses `replicationSendNewlineToPrimary` like you said. The `connSetWriteHandler(conn, NULL); ` that I linked above does not delete the connection itself, only the write handler used for async communication. But `replicationSendNewlineToPrimary` uses `connWrite()` which is a direct write attempt that doesn't require an event handler.

Reply:
So, I am not referring to the write handler. I am referring to the write in 2 threads. this could break some implementations which can have issues supporting read and write being done in 2 different threads. also I am not sure this is the loading phase (maybe I was wrong about referring to replicationSendNewlineToPrimary). also the replicationSendNewlineToPrimary will NOT send '\n' when the repl_transfer_s is NULL AFAIK so the replica will NOT KA when loading (which again is not the case right?) 
I do think that in the RDB_TRANSFER pahse the replica will send occasional ACK using replicationSendAck in the replicationCron, but again I have some concerns about it breaking external TLS or other transport implementations

Reply:
O.K after we discussed and I looked, the replica does not need to write anything to the primary during transfer (loading is somewhat different). I guess it is fine then.